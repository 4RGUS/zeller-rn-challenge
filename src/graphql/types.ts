export interface TableStringFilterInput {
  ne?: string;
  eq?: string;
  le?: string;
  lt?: string;
  ge?: string;
  gt?: string;
  contains?: string;
  notContains?: string;
  between?: string[];
  beginsWith?: string;
}

export interface TableZellerCustomerFilterInput {
  id?: TableStringFilterInput;
  name?: TableStringFilterInput;
  email?: TableStringFilterInput;
  role?: TableStringFilterInput;
}

export interface ZellerCustomer {
  id: string;
  name?: string;
  email?: string;
  role?: string;
}

export interface ZellerCustomerConnection {
  items?: ZellerCustomer[];
  nextToken?: string;
}

export interface ListZellerCustomersQueryVariables {
  filter?: TableZellerCustomerFilterInput;
  limit?: number;
  nextToken?: string;
}

export interface ListZellerCustomersQuery {
  listZellerCustomers?: ZellerCustomerConnection;
}

export interface GetZellerCustomerQueryVariables {
  id: string;
}

export interface GetZellerCustomerQuery {
  getZellerCustomer?: ZellerCustomer;
}

export enum Roles {
  ADMIN = "ADMIN",
  MANAGER = "MANAGER",
}
