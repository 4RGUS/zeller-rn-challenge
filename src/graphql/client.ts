import {
  ApolloClient,
  ApolloLink,
  HttpLink,
  InMemoryCache,
} from "@apollo/client";
import awsconfig from "./aws-exports";
import { AuthOptions, createAuthLink } from "aws-appsync-auth-link";
import { createSubscriptionHandshakeLink } from "aws-appsync-subscription-link";

const auth: AuthOptions = {
  type: "API_KEY",
  apiKey: awsconfig.aws_appsync_apiKey,
};

const httpLink = new HttpLink({
  uri: awsconfig.aws_appsync_graphqlEndpoint,
});

const link = ApolloLink.from([
  createAuthLink({
    url: awsconfig.aws_appsync_graphqlEndpoint,
    region: awsconfig.aws_appsync_region,
    auth,
  }),
  createSubscriptionHandshakeLink(
    {
      url: awsconfig.aws_appsync_graphqlEndpoint,
      region: awsconfig.aws_appsync_region,
      auth,
    },
    httpLink
  ),
]);

const client = new ApolloClient({
  link,
  cache: new InMemoryCache(),
});

export default client;
