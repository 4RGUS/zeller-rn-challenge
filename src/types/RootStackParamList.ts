import { NativeStackNavigationProp } from "@react-navigation/native-stack";

export type RootStackParamList = {
  Home: undefined;
  UserList: undefined;
};

export type UserListScreenNavigationProp = NativeStackNavigationProp<
  RootStackParamList,
  "UserList"
>;
