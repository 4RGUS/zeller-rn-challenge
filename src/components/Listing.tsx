import React from 'react';
import { FlatList, View, Text, StyleSheet } from 'react-native';
import { Roles, ZellerCustomer } from '../graphql/types';
import { mockZellerUsers } from '../mocks/users';

type ListingProps = {
    data: ZellerCustomer[],
    roleType: Roles
}

export const Listing = (props: ListingProps) => {
    const { data, roleType } = props
    const mockedData = mockZellerUsers.filter((value) => value.role === roleType)
    return (
        <FlatList
            data={data.length > 0 ? data : mockedData}
            keyExtractor={(item) => item.id}
            renderItem={({ item }) => (
                <View style={style.container}>
                    <View style={style.initialsBox}>
                        <Text style={style.nameInitialTextStyle}>{item.name?.[0]}</Text>
                    </View>
                    <View
                        style={style.nameContainer}
                    >
                        <Text style={style.nameTextStyle}>{item.name}</Text>
                        <Text style={style.roleTextStyle}>{item.role}</Text>
                    </View>
                </View>
            )}
        />
    )
}

const style = StyleSheet.create({
    container: {
        flexDirection: 'row',
        padding: 12,
        borderRadius: 8,
        marginBottom: 8,
        gap: 12,
        alignItems: "center",
        paddingLeft: 0,
    },
    initialsBox: {
        height: 45,
        width: 45,
        borderRadius: 5,
        backgroundColor: "#E6F0FF",
        justifyContent: "center",
        alignItems: "center"
    },
    nameContainer: {
        width: "100%",
    },
    nameInitialTextStyle: {
        fontSize: 18,
        color: "#4D94FF"
    },
    nameTextStyle: {
        fontWeight: "bold",
    },
    roleTextStyle: {
        color: "#A9A9A9"
    }
})