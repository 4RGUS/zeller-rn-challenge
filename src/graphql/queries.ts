import { gql } from "@apollo/client";

export const LIST_ZELLER_CUSTOMERS = gql`
  query ListZellerCustomers($filter: TableZellerCustomerFilterInput) {
    listZellerCustomers(filter: $filter) {
      items {
        id
        name
        email
        role
      }
      nextToken
    }
  }
`;

export const SEARCH_ZELLER_CUSTOMERS = gql`
  query SearchZellerCustomers(
    $nameFilter: TableStringFilterInput
    $limit: Int
    $nextToken: String
  ) {
    listZellerCustomers(
      filter: { name: $nameFilter }
      limit: $limit
      nextToken: $nextToken
    ) {
      items {
        id
        name
        email
        role
      }
      nextToken
    }
  }
`;
