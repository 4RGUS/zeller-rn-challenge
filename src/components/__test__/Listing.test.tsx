import React from "react";
import { Roles, ZellerCustomer } from "../../graphql/types";
import { render } from "@testing-library/react-native";
import { Listing } from "../Listing";

describe("Listing component", () => {
  const mockData: ZellerCustomer[] = [
    { id: "1", name: "John Doe", role: Roles.ADMIN },
    { id: "2", name: "Jane Smith", role: Roles.MANAGER },
  ];

  it("matches snapshot with provided data", () => {
    const { toJSON } = render(
      <Listing data={mockData} roleType={Roles.ADMIN} />
    );
    expect(toJSON()).toMatchSnapshot();
  });

  it("matches snapshot with empty data (uses mockedData)", () => {
    const { toJSON } = render(<Listing data={[]} roleType={Roles.MANAGER} />);
    expect(toJSON()).toMatchSnapshot();
  });
});
