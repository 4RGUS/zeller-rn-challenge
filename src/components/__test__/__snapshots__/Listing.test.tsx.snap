// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Listing component matches snapshot with empty data (uses mockedData) 1`] = `
<RCTScrollView
  data={
    [
      {
        "email": "karen.stewart@example.com",
        "id": "manager-1",
        "name": "Karen Stewart",
        "role": "Manager",
      },
      {
        "email": "liam.price@example.com",
        "id": "manager-2",
        "name": "Liam Price",
        "role": "Manager",
      },
      {
        "email": "mia.howard@example.com",
        "id": "manager-3",
        "name": "Mia Howard",
        "role": "Manager",
      },
      {
        "email": "noah.scott@example.com",
        "id": "manager-4",
        "name": "Noah Scott",
        "role": "Manager",
      },
      {
        "email": "olivia.turner@example.com",
        "id": "manager-5",
        "name": "Olivia Turner",
        "role": "Manager",
      },
      {
        "email": "pauline.carter@example.com",
        "id": "manager-6",
        "name": "Pauline Carter",
        "role": "Manager",
      },
      {
        "email": "quinn.bailey@example.com",
        "id": "manager-7",
        "name": "Quinn Bailey",
        "role": "Manager",
      },
      {
        "email": "ryan.cooper@example.com",
        "id": "manager-8",
        "name": "Ryan Cooper",
        "role": "Manager",
      },
      {
        "email": "sophia.morgan@example.com",
        "id": "manager-9",
        "name": "Sophia Morgan",
        "role": "Manager",
      },
      {
        "email": "thomas.bennett@example.com",
        "id": "manager-10",
        "name": "Thomas Bennett",
        "role": "Manager",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            K
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Karen Stewart
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            L
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Liam Price
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            M
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Mia Howard
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            N
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Noah Scott
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            O
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Olivia Turner
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            P
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Pauline Carter
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            Q
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Quinn Bailey
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            R
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Ryan Cooper
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            S
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Sophia Morgan
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            T
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Thomas Bennett
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Listing component matches snapshot with provided data 1`] = `
<RCTScrollView
  data={
    [
      {
        "id": "1",
        "name": "John Doe",
        "role": "Admin",
      },
      {
        "id": "2",
        "name": "Jane Smith",
        "role": "Manager",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            J
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            John Doe
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Admin
          </Text>
        </View>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          {
            "alignItems": "center",
            "borderRadius": 8,
            "flexDirection": "row",
            "gap": 12,
            "marginBottom": 8,
            "padding": 12,
            "paddingLeft": 0,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#E6F0FF",
              "borderRadius": 5,
              "height": 45,
              "justifyContent": "center",
              "width": 45,
            }
          }
        >
          <Text
            style={
              {
                "color": "#4D94FF",
                "fontSize": 18,
              }
            }
          >
            J
          </Text>
        </View>
        <View
          style={
            {
              "width": "100%",
            }
          }
        >
          <Text
            style={
              {
                "fontWeight": "bold",
              }
            }
          >
            Jane Smith
          </Text>
          <Text
            style={
              {
                "color": "#A9A9A9",
              }
            }
          >
            Manager
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
